### PR TITLE
[NPU][A5] Supoort MXFP4-W4A4 for Ascend NPU

### PR DESCRIPTION
**Pull request by LinyuanLi0046**
## Motivation

Add Ascend NPU support for ModelSlim W4A4_MXFP4 path of sglang .

## Modifications

- Extend ModelSlimConfig to recognize W4A4_MXFP4 for both linear and MoE layers in the existing modelslim scheme dispatch path.
- Add new ModelSlim W4A4_MXFP4 schemes:
-  W4A4_MXFP4 linear layers
-  W4A4_MXFP4 MOE layers

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** -- add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** -- `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py [hunk 1]
```diff
@@ -141,3 +142,48 @@ def apply(
             bias=bias,
             output_dtype=original_dtype,
         )
+
+
+class NPUW4A4MxFp4LinearMethod(_NPULinearMethodBase):
+
+    def process_weights_after_loading(self, layer: torch.nn.Module):
+        layer.weight.data = torch_npu.npu_format_cast(layer.weight.data, 29).transpose(
```
Comment (LinyuanLi0046) [inline python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py:150]: Several tensors are transposed and then passed forward without being made contiguous. Since transpose() often returns a non-contiguous view, this may introduce implicit copies or fail on NPU kernels with stricter layout requirements.

Comment (LinyuanLi0046) [inline python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py:150]: The code uses npu_format_cast(..., 29) as a magic number and relies on implicit dtype/layout interpretation.

File: python/sglang/srt/layers/quantization/modelslim/modelslim.py [hunk 2]
```diff
@@ -16,6 +16,8 @@
 from sglang.srt.layers.quantization.modelslim.schemes import (
     ModelSlimW4A4Int4,
     ModelSlimW4A4Int4MoE,
+    ModelSlimW4A4MxFp4,
+    ModelSlimW4A4MxFp4MoE,
```
Comment (LinyuanLi0046) [inline python/sglang/srt/layers/quantization/modelslim/modelslim.py:20]: The documentation/support matrix has not been updated. The code registers a new W4A4_MXFP4 dispatch key, but the public docs do not explain the SRT/LLM usage, checkpoint schema, or minimum software requirements.

File: python/sglang/srt/hardware_backend/npu/quantization/fused_moe_method_npu.py [hunk 3]
```diff
@@ -100,6 +101,190 @@ def npu_fused_experts_w4a4(
     return final_hidden_states
 
 
+def mxfp4_gmm_npu(
```
Comment (LinyuanLi0046) [inline python/sglang/srt/hardware_backend/npu/quantization/fused_moe_method_npu.py:104]: The code duplication is high, especially between the prefill/decode routing skeleton and the two-layer GMM + activation flow. This makes future changes to round mode, layout handling, exception handling, or capability guards easy to apply only partially. Please extract a shared helper such as run_mxfp4_moe_block() and separate routing, GMM execution, and merge/finalization logic.

File: python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py (hunk 1 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py:150]: can’t .contiguous() due to kernel's restrict

File: python/sglang/srt/layers/quantization/modelslim/modelslim.py (hunk 2 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/layers/quantization/modelslim/modelslim.py:20]: doc will be updated later

File: python/sglang/srt/hardware_backend/npu/quantization/fused_moe_method_npu.py (hunk 3 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/fused_moe_method_npu.py:104]: can refactor later

File: python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py (hunk 1 above)
Comment (LinyuanLi0046) [reply python/sglang/srt/hardware_backend/npu/quantization/linear_method_npu.py:150]: may add a helper func later
